docs/md_merge.py

```diff
@@ -0,0 +1,357 @@
+import argparse
+import os
+import re
+import shutil
+from argparse import BooleanOptionalAction
+from collections import OrderedDict, defaultdict
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional
+
+MD_HEADER_REGEX = (
+    r"^"            # start of line
+    r"(\s{0,3})"    # Group 1: zero to three whitespaces, tabs etc (four and more spaces are no heading)
+    r"(#{1,6})"     # Group 2: one or up to 6 '#'
+    r"(\s{0,})"     # Group 3: zero or more whitespaces
+    r"([^#].*)"     # Group 4: one or more chars but not starting with another #         
+    r"$"            # end of line
+)
+
+MD_IMAGE_REGEX = (
+    r"^"            # start of line
+    r"(.*)"         # Group 1: optional preceding text
+    r"(!\[)"        # Group 2: "!["
+    r"(.*)"         # Group 3: optional alt text
+    r"(\]\()"       # Group 4: "]("
+    r"([^\s]*)"     # Group 5: image link
+    r"\s*(.*)"      # Group 6: nothing or a whitespace and the optional image title, including quotes
+    r"(\))"         # Group 7: closing parenthesis
+    r'(.*)'         # Group 8: optional subsequent text
+    r"$"            # end of line
+)
+
+MD_LINK_REGEX = (
+    r"^"                        # start of line
+    r"(.*)"                     # Group 1: optional preceding text
+    r"(\[)"                     # Group 2: "["
+    r"(.*)"                     # Group 3: link text
+    r"(\]\()"                   # Group 4: "]("
+    r"([a-zA-Z0-9.:\/#\-_]+)"   # Group 5: link
+    r"(\))"                     # Group 6: closing parenthesis
+    r"(.*)"                     # Group 7: optional subsequent text
+    r"$"                        # end of line
+)
+
+
+@dataclass
+class HeaderLine:
+    depth: int
+    text: str
+
+    def to_text(self, new_depth: Optional[int] = None):
+        hashes = "#" * new_depth if new_depth is not None else "#" * self.depth
+        return f"{hashes} {self.text}"
+
+
+@dataclass
+class LinkLine:
+    link: str
+    link_text: str = ""
+    preceding_text: str = ""
+    subsequent_text: str = ""
+
+    @property
+    def is_external(self):
+        return self.link.startswith("http")
+
+    def to_text(self):
+        raise NotImplementedError()
+
+
+@dataclass
+class ImageLine(LinkLine):
+    title_text: str = ""
+
+    def to_text(self):
+        title_part = f' "{self.title_text}"' if self.title_text else ""
+        return f'{self.preceding_text}![{self.link_text}]({self.link}{title_part}){self.subsequent_text}'
+
+
+@dataclass
+class TextLinkLine(LinkLine):
+
+    def to_text(self):
+        return f"{self.preceding_text}[{self.link_text}]({self.link}){self.subsequent_text}"
+
+    @property
+    def is_anchor_link(self):
+        return self.link.startswith("#")
+
+
+def parse_line(line: str):
+    header_search = re.search(MD_HEADER_REGEX, line, re.IGNORECASE)
```

Comment:
Nit: Although the function name seems to be obvious, here a doctoring describing parsing strategy, assumed inputs, and polymorphic outputs might add value

Reply:
I added some more docstrings and type hints.